### ISSUE TITLE
Shift in collected observations by one timestep

### ISSUE THREAD

**Issue by antoine-galataud** [IBM/rl-testbed-for-energyplus#64]
Found this problem while on a model where HVAC system shuts off at night and with `CurrentTime` EnergyPlus built-in variable used as observation.

### Problem description

On this model, I have HVAC starting at 6:00 AM. Current time is reported as an observation as fractional time. On a given episode, I get this following wrong trajectory reported

tmp1 | tmp2 | tmp3 | time | remaining ts | rew | action
-- | -- | -- | -- | -- | -- | --
16.0 | 11.3 | 5.3 | 6.25 | 51.0 | 1.000000 | 20.0
16.1 | 15.7 | 15.6 | 6.50 | 50.0 | 0.373034 | 23.5

What happens:

The trajectory we feed to the agent is the following:

```
ts 1 (e.g 6:15), obs vector at ts -2 (before HVAC on), rew for obs at ts -2, default action
ts 2 (e.g 6:30), obs vector at ts -2 (default action), rew for obs at ts -2, act 1, 
ts 3 (e.g 6:45), obs vector at ts -2 (act 1), rew for obs at ts -2, act 2
...
```

This is a shift of 1 timestep for the observations sent to the agent and for computing the reward.

This can be confirmed by asking EnergyPlus to output results in CSV or SQL:

time | tmp1 | tmp2 | tmp3 | tmp4 | tmp5
-- | -- | -- | -- | -- | --
2020-01-01 06:00:00 | 7.740999 | 7.270150 | 15.277604 | 4.518456 | 6.046737
2020-01-01 06:15:00 | 15.487861 | 15.406199 | 16.012610 | 15.586937 | 15.586152

Here we can see that there is no such shift (indoor temperatures have increased already at 06:15).

The correct trajectory should be:

```
ts 1 (e.g 6:15), obs vector at ts -1 (default action), rew for obs at ts -1, default action
ts 2 (e.g 6:30), obs vector at ts -1 (act 1), rew for obs at ts -1, act 1
ts 3 (e.g 6:45), obs vector at ts -1 (act 2), rew for obs at ts -1, act 2
```

### Reason

I believe the incorrect calling point for EMS program is used. The one currently used is `AfterPredictorAfterHVACManagers`. As stated in https://bigladdersoftware.com/epx/docs/8-2/input-output-reference/group-energy-management-system-ems.html#energymanagementsystemprogramcallingmanager and https://bigladdersoftware.com/epx/docs/8-2/ems-application-guide/ems-calling-points.html#ems-calling-points, this calling point happens **before HVAC calculations**, so variables and meters values are still the one from previous time step.

### Solution

Apply action(s) at `AfterPredictorAfterHVACManagers` calling point, but collect observations after zone time step is done, e.g. at `EndOfZoneTimestepAfterZoneReporting` calling point

**Comment by antoine-galataud** [comment]
@takaomoriyama please let me know your thoughts on this. I have tested the solution on one of my environments, and it does fix the reported issue. Performance (episode reward mean) is marginally improved, but convergence happens faster (tested on 4 trials).

**Comment by antoine-galataud** [comment]
I added an example of EnergyPlus model change to fix this problem. See #65

**Comment by antoine-galataud** [comment]
@ZHANG-QINGANG Hi!
Thank you for the detailed issue report.

At first sight, this looks like a slightly different problem than the original one: in the original issue there's a gap of one EnergyPlus time step between action decided by agent and observations collected (observations collected and reward computed on t-2 instead of t-1).
In your report, the shift seems to come from within the same observation, but actions and observations seem well aligned, assuming that your excel spreadsheet:
- logs on the same row action and observation resulting from this action, and not observation before action is applied).
- was generated using a single program with `AfterPredictorAfterHVACManagers` calling point only

See below for illustration:
![https___user-images githubusercontent com_53898971-mh](https://user-images.githubusercontent.com/5330926/155959030-50535d8c-8bf8-4626-9432-c0dd0b431547.png)

Then I'd say that your solution should come as a complement.

The question I have is about the tests you did with the proposed solution for the original issue, which consists in using `EndOfZoneTimestepAfterZoneReporting` for observations collection, and the fact that it results in 0W of power sent by EnergyPlus. Collecting observations near the end of the timestep should positively impact zones-related calculations, like indoor temperatures and meters. This is what I observe in my projects (they use HVAC power mainly), this is also stated in [E+ documentation](https://bigladdersoftware.com/epx/docs/9-5/input-output-reference/group-energy-management-system-ems.html#field-energyplus-model-calling-point):

> EndOfZoneTimestepAfterZoneReporting. This calling point happens each zone timestep just after the output reports are updated for zone-timestep variables and meters. This calling point is the last one of a timestep and is useful for making control decisions for the next zone timestep using the final meter values for the current zone timestep.

Could you please share your IDF file, or at least the EMS part?

**Comment by ZHANG-QINGANG** [comment]
Hi @antoine-galataud,

I attached my modified IDF file. It follows the structure of the 2 zone model provided by this testbed. It should be noted that I generated the excel sheet by using the following pseudo-code, not the reported CSV file of E+. The following figure is the code I used to record data. Then, in the same row, the temperature terms are from "o2". I think it is reasonable since the temperature of "o2" is the result of executing "a"  at "o". What do you think?

I tested your proposed solution by directly running the IDF model you modified. I got "0" when I try to record the observations. I feel the "0" is a little weird. But I have not found the reason so far.

For my IDF file, I added several sensors to collect the power of Equipment. I also wrote a program named "CalculatePredictor" to calculate the cooling load.

![image](https://user-images.githubusercontent.com/53898971/155966089-a757646b-0c91-4370-83bd-81f5459b00e2.png)

> Reset the environment and get observation _o_
> **for** _i_ in range(_n_), **run**:
      > Generate action _a_, according to _o_
      > Execute _a_ and obtain new observation _o2_
      >  Store _(o,a,o2)_
      > _o←o_2_

[2ZoneDataCenterHVAC_wEconomizer_Temp_Fan_ZhangQingang.zip](https://github.com/IBM/rl-testbed-for-energyplus/files/8152393/2ZoneDataCenterHVAC_wEconomizer_Temp_Fan_ZhangQingang.zip)

**Comment by antoine-galataud** [comment]
Thank you @ZHANG-QINGANG for sharing your IDF file. This allowed to find a problem in the original fix I provided. The statement:
```
SET tmp_val1 = @ExtCtrlAct 0 7
```
should be in the `ExtCtrlBasedSetpointManager` program, and not in `ExtCtrlBasedObservationCollector`.

Find attached your file with a fix. Let me know if that works for you:
- do you get non 0 power readings?
- does it help with aligning CPU utilization and ITE Power?

Thanks

[2ZoneDataCenterHVAC_wEconomizer_Temp_Fan_ZhangQingang.idf.zip](https://github.com/IBM/rl-testbed-for-energyplus/files/8154418/2ZoneDataCenterHVAC_wEconomizer_Temp_Fan_ZhangQingang.idf.zip)

**Comment by ZHANG-QINGANG** [comment]
Hi @antoine-galataud ,

I tested you new solution. The "0" reading issue was solved and the CPU-ITEPower was aligned.

However, it seems the zone temperature still cannot been well controlled, as illustrated in the following figure.

![image](https://user-images.githubusercontent.com/53898971/156016839-cdb71056-8c04-4521-ad03-4a08785b6134.png)

I guess, even though the observation was correct, the corresponding actions will be executed in the next time step.

![image](https://user-images.githubusercontent.com/53898971/156017292-47d515e3-06fa-4517-9edb-3250ebbcc5d5.png)

What do you think?

**Comment by antoine-galataud** [comment]
@ZHANG-QINGANG I would say now that the alignment problem is solved, but the policy learned is sub-optimal as it is unable to anticipate on CPU utilization increase and thus temperature increase in zones. But this is a somewhat different problem.

Now the alignment seems correct and we observe in your last spreadsheet that when CPU load / temperatures increase, the subsequent action taken as a result is higher.
For instance:
- at row 9, decided action is 3.82 kg/s since CPU util. is 0.5 and zones temperatures are < 25.0C
- at row 10, decided action is 5.11 kg/s since CPU util. is now higher, as well as zones temperatures.

**Comment by ZHANG-QINGANG** [comment]
@antoine-galataud  
Yes, I agree with your opinion: **it is unable to anticipate on CPU utilization increase and/or temperature increase in zones.** Your solution is good.

My only concern is if the working process follows the figure in my last message, It might not follow the control logic of E+: **Predict Equipment Load~Adjust HVAC Manager accordingly**. Then, the control will be quite passive. Problems may arise when the utilization changes frequently and drastically. Here I attached a case where utilization changes from 0.2 to 0.8. Of course, this can remain as future research. 

It is a pleasure to discuss with you, very helpful. :-)

![image](https://user-images.githubusercontent.com/53898971/156090396-f90f090c-f58b-40e4-8e66-8abe044d2efe.png)

**Comment by antoine-galataud** [comment]
@takaomoriyama the problem is now fixed with the new fix I pushed in #77. I suggest we re-close this issue, unless you think otherwise.

**Comment by takaomoriyama** [comment]
@antoine-galataud OK. That's great.
If we need some more discussion, let's open another issue.

@ZHANG-QINGANG Thank you for your contribution.
